loadtest/control/browsercontroller/controller_test.go

```diff
@@ -0,0 +1,435 @@
+)
+
+func TestNew(t *testing.T) {
+	store, err := memstore.New(nil)
+	require.NoError(t, err)
+	user := userentity.New(userentity.Setup{Store: store}, userentity.Config{
+		ServerURL:    "http://localhost:8065",
+		WebSocketURL: "ws://localhost:8065",
+		Username:     "testuser",
+		Email:        "test@example.com",
+		Password:     "testpass",
+	})
+
+	statusChan := make(chan control.UserStatus, 10)
+	defer close(statusChan)
+
+	t.Run("controller is created successfully", func(t *testing.T) {
+		controller, err := New(1, user, statusChan)
+		require.NoError(t, err)
+		require.NotNil(t, controller)
+		require.Equal(t, 1, controller.id)
+		require.Equal(t, user, controller.user)
+		require.Equal(t, LT_BROWSER_API_URL, controller.ltBrowserApiUrl)
+		require.Equal(t, 1.0, controller.rate)
+		require.NotNil(t, controller.httpClient)
+	})
+
+	t.Run("controller is not created if user is nil", func(t *testing.T) {
+		controller, err := New(1, nil, statusChan)
+		require.Error(t, err)
+		require.Nil(t, controller)
+		require.Contains(t, err.Error(), "user cannot be nil")
+	})
+
+	t.Run("controller is not created if status channel is nil", func(t *testing.T) {
+		controller, err := New(1, user, nil)
+		require.Error(t, err)
+		require.Nil(t, controller)
+		require.Contains(t, err.Error(), "status channel cannot be nil")
+	})
+}
+
+func newController(t *testing.T) (*BrowserController, chan control.UserStatus) {
+	t.Helper()
+
+	store, err := memstore.New(nil)
+	require.NoError(t, err)
+
+	user := userentity.New(userentity.Setup{Store: store}, userentity.Config{
+		ServerURL:    "http://localhost:8065",
+		WebSocketURL: "ws://localhost:8065",
+		Username:     "testuser",
+		Email:        "test@example.com",
+		Password:     "testpass",
+	})
+
+	statusChan := make(chan control.UserStatus, 10)
+
+	controller, err := New(1, user, statusChan)
+	require.NoError(t, err)
+	require.NotNil(t, controller)
+
+	return controller, statusChan
+}
+
+func TestRun(t *testing.T) {
+	t.Run("controller run fails if user is nil", func(t *testing.T) {
+		statusChanWithoutUser := make(chan control.UserStatus, 10)
+		defer close(statusChanWithoutUser)
+
+		controllerWithEmptyUser := &BrowserController{
+			id:              1,
+			user:            nil,
+			status:          statusChanWithoutUser,
+			rate:            1.0,
+			stopChan:        make(chan struct{}),
+			stoppedChan:     make(chan struct{}),
+			ltBrowserApiUrl: LT_BROWSER_API_URL,
+		}
+
+		go controllerWithEmptyUser.Run()
+		statusF := <-statusChanWithoutUser
+		require.Error(t, statusF.Err)
+		require.Equal(t, control.USER_STATUS_FAILED, statusF.Code)
+		require.Contains(t, statusF.Err.Error(), "browser controller was not initialized")
+	})
+
+	t.Run("controller lifecycle with browser API not running", func(t *testing.T) {
+		controllerWithUser, statusChan := newController(t)
+
+		go controllerWithUser.Run()
+
+		// Test that controller is running after Run is called
+		status := <-statusChan
```

Comment:
Should we add a `select` with a timeout in case the code fails, so that the test can finish in a reasonable amount of time? This comment applies to other channels being read in this file.

Reply:
> in case the code fails,

Can you clarify which code would fail?

> the test can finish in a reasonable amount

Simulation test are supposed to be unbounded and run till infinity